### PR TITLE
Activity adapter v2

### PR DESCRIPTION
**Pull request by SaurabhCompro**

### CODE REVIEW

File: lib/config/index.js [hunk 1]
```diff
@@ -41,7 +41,12 @@ exports.AUTH_API_URLS = {
 };
 
 exports.ACTIVITY_API_URLS = {
-    //TODO
+    "newAttemptAPI": "/{orgid}/{productId}/activity/{activityId}/start",
+    "activityDetailsAPI": "/{orgid}/{productId}/activity/{activityId}",
+    "existingAttemptAPI": "/{orgid}/{productId}/activity/{activityId}/attempt/{attemptId}",
+    "userresponseAPI": "/{orgid}/{productId}/activity/{activityId}/attempt/{attemptId}/response",
```
Comment (deshbirc) [inline lib/config/index.js]: Wrong case should `userResponseAPI`

File: lib/config/index.js [hunk 2]
```diff
@@ -53,7 +58,8 @@ exports.COLLAB_API_URLS = {
 };
 
 exports.PRODUCT_API_URLS = {
-    //TODO
+    "questionsAPI": "/{orgid}/products/{productId}/activities/{activityId}",
```
Comment (deshbirc) [inline lib/config/index.js]: Should this be `ActivityAPI` ?

File: lib/helpers/lib/api/index.js [hunk 3]
```diff
@@ -90,7 +90,7 @@ function genericAPICaller(apiName, method, url, params) {
 };
 
 //This function constructs the url of comproDLS API.
-function constructAPIUrl(url,orgid,userId,productId) {
+function constructAPIUrl(url,orgid,userId,productId, activityId, attemptId, questionId) {
```
Comment (deshbirc) [inline lib/helpers/lib/api/index.js]: Please a comment, clarifying which parameters are optional

File: lib/services/activity/activity.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+function Activity(productId, activityId) {
+    //When Activity constructor is called, check for instance type
+    //comproDLS.Activity() will give instance type as ComproDLS
+    //So the control goes to else
+    if (this instanceof Activity) {
+        this.productId = productId; //Here this is instance of Activity
+        this.activityId = activityId; //Here this is instance of Activity
+    } else { //New object of Activity is created and its instance type is Activity
+        var act = new Activity(productId, activityId);
```
Comment (deshbirc) [inline lib/services/activity/activity.js]: Great comments, thanks!

File: lib/services/activity/activity.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            dfd.reject(err);
+        } else {
+
+    //Passed all validations, Contruct API url
+    var url = config.DEFAULT_HOSTS['ACTIVITY'] + config.ACTIVITY_API_URLS.activityDetailsAPI;
+
+    url = helpers.api.constructAPIUrl(url, self.orgId, null, self.productId, self.activityId);
+
+    //Setup request with URL and Params
+    var requestAPI = request.get(url);
```
Comment (deshbirc) [inline lib/services/activity/activity.js]: What do you mean by "requestAPI"? I'm not aware of any such API

File: lib/services/activity/activity.js [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    requestAPI = helpers.api.setupAPIToken(requestAPI, self.token);
+
+    //Call GET activity details Api
+    requestAPI.end(function (err, response) {
+        if (err) {
+
+            err = new DLSError(helpers.errors.ERROR_TYPES.API_ERROR, err);
+            dfd.reject(err);
+        } else {
+             console.log(response.body);
```
Comment (deshbirc) [inline lib/services/activity/activity.js]: Why do we need to log this? Is it a debug call?

File: lib/services/activity/activity.js [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+Activity.prototype.getAttempt = function (attemptId) {
+    var newAttempt = new attempt(this.productId, this.activityId, attemptId);
+    newAttempt.setAuthProperties(this.orgId, this.token);
+    return newAttempt.resume();
+}
+
+Activity.prototype.newAttempt = function (attemptId) {
+   
+    var newAttempt = new attempt(this.productId, this.activityId, null);
+     console.log("$$$$$$$$$$$$");
```
Comment (deshbirc) [inline lib/services/activity/activity.js]: What is this???

File: lib/services/activity/activity.js [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    //Call GET activity details Api
+    requestAPI.end(function (err, response) {
+        if (err) {
+
+            err = new DLSError(helpers.errors.ERROR_TYPES.API_ERROR, err);
+            dfd.reject(err);
+        } else {
+             console.log(response.body);
+
+            var responseObject = response.body;
```
Comment (deshbirc) [inline lib/services/activity/activity.js]: lets call this `var reponseBody`

File: lib/services/activity/attempt.js [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    }
+
+    function setProperties(authProperties) {
+       
+        this.authProperties = authProperties;
+        
+    }
+    
+        //Privilaged Methods
+    Attempt.prototype.getAuthProperties = function () {
```
Comment (deshbirc) [inline lib/services/activity/attempt.js]: Why do we need this function? Who calls it?

File: lib/services/activity/attempt.js [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+/*********************************
+ * Constructor
+ **********************************/
+
+function Attempt(productId, activityId, attemptId, authProperties) {
+    this.productId = productId;
+    this.activityId = activityId;
+    this.attemptId = attemptId;
+    
+    var authProperties;
```
Comment (deshbirc) [inline lib/services/activity/attempt.js]: it should be called `authCredentials`

File: lib/services/activity/attempt.js [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        this.authProperties = authProperties;
+        
+    }
+    
+        //Privilaged Methods
+    Attempt.prototype.getAuthProperties = function () {
+        return getProperties();
+    };
+    
+    Attempt.prototype.setAuthProperties = function (orgid, token) {
```
Comment (deshbirc) [inline lib/services/activity/attempt.js]: Why do we need this? 

In any case call it `setAuthCredentials`

File: package.json [hunk 12]
```diff
@@ -15,7 +15,8 @@
     "q": "~1.4.1",
     "superagent": "~1.7.2",
     "validate.js": "^0.9.0",
-    "socket.io-client": "^1.4.5"
+    "socket.io-client": "^1.4.5",
```
Comment (deshbirc) [inline package.json]: Lets make this dependency specific - socket.io can change tomorrow and break everything

File: lib/config/index.js (hunk 2 above)
Comment (SaurabhCompro) [inline lib/config/index.js]: Ok

File: lib/services/activity/activity.js (hunk 5 above)
Comment (SaurabhCompro) [inline lib/services/activity/activity.js]: It is just a variable which stores reference to super agent request object. Copied from Auth index.js

File: README.md [hunk 13]
```diff
@@ -212,7 +212,94 @@ auth.getUserProfile().then(
...
+Following is an example of using ACTIVITY API SDK adaptor to get profile information for the user. In this function, the SDK adaptor abstracts the knowledge of underlying REST URLs & Parameters, and offers a simpler interface. For a list of all SDK ACTIVITY AUTH functions, see [ACTIVITY Adaptor Documentation](https://github.com/comprodls/comprodls-sdk-js/wiki/09_ACTIVITY-Adaptor).
+
+```javascript
+var activity = comproDLS.Activity(productId, activityId);
+
+/* Get all the attempts of an activity for a user */
+activity.getDetails().then(
+  function success(response) {
+    <!--Activity {
+       activityid (string),
```
Comment (deshbirc) [inline README.md]: The activity json should available as `response.activity`. Please apply this rule everywhere -- do NOT use the ROOT response object

File: README.md [hunk 14]
```diff
@@ -212,7 +212,94 @@ auth.getUserProfile().then(
...
+  }, 
+  function error(err) {
+     /* --- Error --- */   
+  }
+);
 
+// Save user response
+attempt.saveData(attempt, savedstate, userresponse).then(
+  function success(response) {
+      console.log("Success " + attempt.id);
```
Comment (deshbirc) [inline README.md]: should be console.log("Success " + response.attempt...);

File: README.md [hunk 15]
```diff
@@ -212,7 +212,94 @@ auth.getUserProfile().then(
...
+     /* --- Error --- */   
+  }
+);
+
+
+// Submit
+
+activity.submitAttempt(attempt, userresponse).then(
+  function success(response) {
+      console.log("Success " + attempt.id);
```
Comment (deshbirc) [inline README.md]: `console.log("Success " + response.attempt.id);`

File: README.md [hunk 16]
```diff
@@ -212,7 +212,94 @@ auth.getUserProfile().then(
...
+
+
+// Get scores
+activity.getAttemptScore().then(
+  function success(response) {
+  
+//get /{orgid}/{productid}/activity/{activityid} 
+//loop and return score for attempt.id
+      console.log("Success " + attempt.id);
+  },
```
Comment (deshbirc) [inline README.md]: ???

File: lib/services/activity/activity.js (hunk 5 above)
Comment (deshbirc) [inline lib/services/activity/activity.js]: `var req = request.get(url);`

File: README.md [hunk 17]
```diff
@@ -212,7 +212,105 @@ auth.getUserProfile().then(
...
+
+```javascript
+var activity = comproDLS.Activity(productId, activityId);
+
+/* Get all the attempts of an activity for a user */
+activity.getDetails().then(
+  function success(response) {
+  
+    var activity = response.activity;
+    <!--Activity {
```
Comment (manavmanocha) [inline README.md]: This is not a valid Javascript comment.

File: lib/services/activity/activity.js [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+    return newAttempt.resume();
+}
+
+Activity.prototype.newAttempt = function (attemptId) {
+    var newAttempt = new attempt(this.productId, this.activityId, null);
+    newAttempt.setAuthCredentials(this.orgId, this.token);
+    return newAttempt.start();
+}
+
+Activity.prototype.getQuestions = function () {
```
Comment (manavmanocha) [inline lib/services/activity/activity.js]: This can be renamed to getAllQuestions for clarity

File: lib/services/activity/attempt.js [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+        } else {
+
+    //Passed all validations, Contruct API url
+    var url = config.DEFAULT_HOSTS['ACTIVITY'] + config.ACTIVITY_API_URLS.newAttemptAPI;
+
+    url = helpers.api.constructAPIUrl(url, orgid, null, self.productId, self.activityId);
+   
+    //Contruct parameters
+
+    var params = {
```
Comment (manavmanocha) [inline lib/services/activity/attempt.js]: I think this should not be hardcoded and passed by caller

Comment (SaurabhCompro) [inline lib/services/activity/attempt.js]: Start parameters can be passed on now.

File: lib/services/activity/activity.js (hunk 18 above)
Comment (SaurabhCompro) [inline lib/services/activity/activity.js]: Renamed the function to getAllQuestions

File: package.json (hunk 12 above)
Comment (manavmanocha) [inline package.json]: Done Sir.

**Pull request merged**